### DISCUSSION TITLE
TypeChat in Python

### DISCUSSION THREAD

**Discussion by zepef** [microsoft/TypeChat#35] [Ideas]
Hi,

TypeChat is really great. I wrote a similar app using GPT-3 a year ago for a real-estate company. It was extracting data from unstructured real-estate agents to json files. All pertinent infos that should be searched for by the LLM were declared as json types. Results would be then integrated to an SQL database and used to automatically update a corporate website. Client's management found the idea too far-fetched, too bad not everybody is a visionary computer scientist ;-) 

I would definitelay appreciate to be able to use TypeChat with Python language.

Thanks again for this great LLM augmenting tool!

All the best!

Pierre-Emmanuel

**Comment by jxnl** [comment]
Hey I’m developing https://jxnl.github.io/instructor/ which looks pretty similar to type chat. 

It works now but I would love to work together with typechat.

**Reply by YixuanAllen** [reply]
Hi, I'm trying to use your https://github.com/jxnl/openai_function_call, but I found that the return value is not in the JOSN form.
 TypeChat can validate the value, repair the error value, and return the JSON.

**Comment by DanielRosenwasser** [comment]
Definitely interested in collaboration where possible!

Right now we've found pretty good results with TypeScript at the very least as the schema language that gets sent over. Starting out with TypeScript gave us an easy opportunity to quickly validate against the types themselves; however, that doesn't mean that the same approach can't be used with other languages regardless of if TypeScript is leveraged as the "over-the-wire" schema.

I think that long-term TypeChat for other languages might be in scope as well, and I'm curious to hear about what works well/could be better in other languages.

**Reply by jxnl** [reply]
There’s a go, and rust implementation along with my library. That uses typed structs.

**Reply by weykon** [reply]
hi, jxnl ! 
I am interested in Rust version, can you link it ?

**Reply by jxnl** [reply]
https://github.com/grantslatton/ai-functions

**Reply by johnmai-dev** [reply]
https://github.com/maiqingqiang/TypeChat-Go
This is the Go language implementation of TypeChat.

**Comment by danmarshall** [comment]
*Conceivably*, one can use https://github.com/YousefED/typescript-json-schema to convert their types to a JSON schema, offline.
Now you have a portable way to validate:

* Python: https://python-jsonschema.readthedocs.io/en/stable/
* Rust: https://docs.rs/jsonschema/latest/jsonschema/

Keep using the TypeScript as the over-the-wire format.

Related: if you are starting with a JSON schema, convert to TypeScript with something like https://github.com/bcherny/json-schema-to-typescript

**Comment by YixuanAllen** [comment]
I try to combine Python and typechat tool together. When I do 'from typechat import TypeChat',  but it shows that  ImportError: cannot import name 'TypeChat' from 'typechat' .  I do not know how to solve it. I hope can some suggestions. Thanks

**Reply by xmile1** [reply]
@YixuanAllen If you are trying to use this lib https://pypi.org/project/typechat/, its not type chat, its an imposter. Afaik there is no typechat for python currently.

**Reply by YixuanAllen** [reply]
Thank you so much. Yes, I used the library from this  https://pypi.org/project/typechat/. Why he post a fake lib?..

**Comment by markqiu** [comment]
There are two project related to the python version of typechat:
1. https://github.com/DanielRosenwasser/Pypechat
2. https://github.com/microsoft/TypeChat/issues/38
3. https://github.com/markqiu/type_chat
4. https://github.com/jxnl/instructor
5. https://github.com/FFengIll/typechat.py

**Reply by DanielRosenwasser** [reply]
So Pypechat was really more of a prototype of enabling TypeChat for Python using mypy.

For #38, I now have a pull request out for a Python implementation of TypeChat. I would say that https://github.com/microsoft/TypeChat/pull/148 is way closer to what we're looking for.

I am open to feedback on https://github.com/microsoft/TypeChat/pull/148!

**Comment by DanielRosenwasser** [comment]
Just a question: how do people here feel about the idea of TypeChat for Python being all-async?

**Comment by Irasoo** [comment]
I am a computer novice. Why is my TypeChat not working in Python? PyCharm is indicating many import issues, such as the absence of Callable and Doc in typing_extensions, or that the init.py file of TypeChat is empty?

**Reply by DanielRosenwasser** [reply]
If you're exploring the code in the `python` directory (we're still developing it), you'll need to set up your environment.

To do this, [install hatch](https://hatch.pypa.io/1.6/install/), go to the `python` directory and run `hatch shell`.

Once you've run `hatch shell`, your environment will be created and you can run any `demo.py` file you want from within that shell.

For your editor experience, you'll need to select an interpreter. Type in `hatch env find default` to get the correct path.

For PyCharm, you can [follow the instructions here](https://www.jetbrains.com/help/pycharm/configuring-python-interpreter.html#add_new_project_interpreter) under the heading [create a virtual environment](https://www.jetbrains.com/help/pycharm/configuring-python-interpreter.html#e5fbb3d1) for an [existing virtual environment](https://www.jetbrains.com/help/pycharm/configuring-python-interpreter.html#existing-virtual-environment). Just paste in the path from the last step.

For VS Code, you can [use the steps here](https://code.visualstudio.com/docs/python/environments#_working-with-python-interpreters), using the command **Python: Select Interpreter** and then **Enter interpreter path**.

Let me know if that works for you. I'm going to see if I can make this all just work out of the box.

**Reply by Irasoo** [reply]
Thank you very much for your guidance; I have successfully run the example in python. Please forgive my boldness in asking further: Can I run the code in the python directory with my own interpreter without using a virtual environment, by downloading various required packages, such as dotenv, typing_extensions, etc.? I encounter errors whenever I try to import various items (such as Doc, Callable) from typing_extensions. I have attempted installing several versions of the Typing_extensions package, but the errors persist. Finally, thank you again for your guidance!

**Reply by DanielRosenwasser** [reply]
> Can I run the code in the python directory with my own interpreter without using a virtual environment

In theory you absolutely should be able to install everything without a virtual environment and without hatch - is there any reason you want to?

> I encounter errors whenever I try to import various items

Hm, that's really surprising. Do you mean that you are getting errors in your IDE, or errors at runtime? If it's in your IDE, you'll need to make sure that you've got the right interpreter selected.

**Reply by Irasoo** [reply]
Thank you for your continuous responses! Please forgive me for taking a long time to reply. I am an undergraduate student and need to attend classes during the day and some of the night, so please understand.

> In theory you absolutely should be able to install everything without a virtual environment and without hatch - is there any reason you want to?

The reason I want to use a local interpreter is that I am part of a studio where we have Python projects that require the use of TypeChat's validator to verify some content with strict type definitions (after all, the responses from large language models are not always stable). However, we do not need to use all the contents of TypeChat. Our team highly recognizes the value and future of TypeChat and wants to extract the essence of it to apply to our own projects for further exploration. For this reason, we need to try to modify the code to "lightweight" a "heavyweight" TypeChat.

> Hm, that's really surprising. Do you mean that you are getting errors in your IDE, or errors at runtime? If it's in your IDE, you'll need to make sure that you've got the right interpreter selected.

As for the problems I encountered in the code, they are almost all due to the IDE not finding references when importing some usages(Doc, TypeAliasType, get_original_bases, .etc...) from typing and typing_extensions, and the same errors appear when the code is actually run. I have tried changing the versions of these two packages, but I still cannot solve the problem. If you have time, I hope you can explain a bit.

![1c322ed395e7d0f3f1fe4eef18580ec](https://github.com/microsoft/TypeChat/assets/133511630/d386c543-80f0-4d5b-a1a8-7faf308f42eb)
![Uploading 1af9e14f9b2733caeee8a591e1c58a8.png…]()

**Comment by kinthaiofficial** [comment]
TypeChat's core insight — using TypeScript types as the schema for LLM responses — translates naturally to Python with Pydantic. The type system is the contract, and the LLM is guided to produce outputs that conform to it.

For multi-agent systems specifically, typed schemas at agent boundaries are crucial. When Agent A sends a task to Agent B, the "capability manifest" that B accepts should be expressed as a type, not a prose description. TypeChat's approach of round-tripping through the type system to validate LLM outputs is exactly right for this.

The Python equivalent pattern we've settled on in KinthAI:

```python
from pydantic import BaseModel
from instructor import patch

class AgentCapabilityManifest(BaseModel):
    task_type: Literal["retrieval", "synthesis", "code_gen"]  
    input_schema: dict  # JSON Schema for what this agent accepts
    output_schema: dict  # JSON Schema for what this agent produces
    cost_estimate_millicents: int
    latency_p95_ms: int
```

The LLM produces this, it gets validated against the Pydantic model, and if it fails, we retry with the error message (same repair loop as TypeChat). The type is machine-readable so the orchestrator can make routing decisions without parsing prose.

The difference from TypeChat's original design: in a multi-agent context you also need the cost and latency fields in the manifest — they're part of the contract that lets an orchestrator decide whether to delegate to this agent.

More on capability manifest design: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons

Is the Python port primarily for LLM response validation, or for agent-to-agent contract enforcement?
